internal/uvm/virtual_device.go

```diff
@@ -117,30 +130,36 @@ func (uvm *UtilityVM) AssignDevice(ctx context.Context, deviceID string) (*VPCID
 		return nil, err
 	}
 	result := &VPCIDevice{
-		vm:               uvm,
-		VMBusGUID:        vmBusGUID,
-		deviceInstanceID: deviceID,
-		refCount:         1,
+		vm:                   uvm,
+		VMBusGUID:            vmBusGUID,
+		deviceInstanceID:     deviceID,
+		virtualFunctionIndex: index,
+		refCount:             1,
 	}
-	uvm.vpciDevices[deviceID] = result
+	uvm.vpciDevices[key] = result
 	return result, nil
 }
 
 // removeDevice removes a vpci device from a uvm when there are
 // no more references to a given VPCIDevice. Otherwise, decrements
 // the reference count of the stored VPCIDevice and returns nil.
-func (uvm *UtilityVM) removeDevice(ctx context.Context, deviceInstanceID string) error {
+func (uvm *UtilityVM) removeDevice(ctx context.Context, deviceInstanceID string, index uint16) error {
+	key := VPCIDeviceKey{
+		deviceInstanceID:     deviceInstanceID,
+		virtualFunctionIndex: index,
+	}
+
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
 
-	vpci := uvm.vpciDevices[deviceInstanceID]
+	vpci := uvm.vpciDevices[key]
 	if vpci == nil {
-		return fmt.Errorf("no device with ID %s is present on the uvm %s", deviceInstanceID, uvm.ID())
+		return fmt.Errorf("no device with ID %s and index %d is present on the uvm %s", deviceInstanceID, index, uvm.ID())
 	}
 
 	vpci.refCount--
 	if vpci.refCount == 0 {
-		delete(uvm.vpciDevices, deviceInstanceID)
+		delete(uvm.vpciDevices, key)
 		return uvm.modify(ctx, &hcsschema.ModifySettingRequest{
 			ResourcePath: fmt.Sprintf(resourcepaths.VirtualPCIResourceFormat, vpci.VMBusGUID),
 			RequestType:  requesttype.Remove,
```

Comment:
Wouldn't hcs need the index to know which to remove? Or is this removing all vfs?

Reply:
Each time you make a vpci assignment, a new vmbus path is created for that device instance. So each vf would have it's own vmbus path. This call to HCS removes the device at the vmbus path.

Reply:
Ok sweet!